### ISSUE TITLE
Cannot configure conversion of value types

### ISSUE THREAD

**Issue by Liversage** [agileobjects/AgileMapper#114]
I wanted to use AgileMapper but I've hit what seems like a brick wall when it comes to [type conversion](https://agilemapper.readthedocs.io/Type-Conversion/). To give some context I would like to explain what I want to do.

To combat [primitive obsession](http://wiki.c2.com/?PrimitiveObsession) I create domain objects that are like integers, strings and GUIDs except they are "strongly typed". These types just wrap a primitive type so they are value types (`struct`) to make them efficient. For instance a "strongly typed" user ID:

```C#
public readonly struct UserId : IEquatable<UserId>
{
	private readonly int value;
	public UserId(int value) => this.value = value;
	public static implicit operator UserId(int value) => new UserId(value);
	public static explicit operator int(UserId playerId) => playerId.value;
	public bool Equals(UserId other) => value.Equals(other.value);
	public override bool Equals(object obj) => obj is UserId other && Equals(other);
	public override int GetHashCode() => value.GetHashCode();
	public override string ToString() => value.ToString();
	public static bool operator ==(UserId playerId1, UserId playerId2) => playerId1.Equals(playerId2);
	public static bool operator !=(UserId playerId1, UserId playerId2) => !(playerId1 == playerId2);
}
```

Here is another one that wraps a reference type (a `string`) which requires the code to be slightly different:

```C#
public readonly struct Nickname : IEquatable<Nickname>
{
	private readonly string value;
	public Nickname(string value) => this.value = value;
	public static implicit operator Nickname(string value) => new Nickname(value);
	public static explicit operator string(Nickname nickname) => nickname.value;
	public bool Equals(Nickname other) => ReferenceEquals(value, other.value) || value != null && value.Equals(other.value);
	public override bool Equals(object obj) => obj is Nickname nickname && Equals(nickname);
	public override int GetHashCode() => (value?.GetHashCode()).GetValueOrDefault();
	public override string ToString() => value?.ToString() ?? string.Empty;
	public static bool operator ==(Nickname nickname1, Nickname nickname2) => nickname1.Equals(nickname2);
	public static bool operator !=(Nickname nickname1, Nickname nickname2) => !(nickname1 == nickname2);
}
```

Using these types in my domain increases the quality of my code. However, I have to interact with external systems, do serialization etc. so I have need to map these types to DTO types that are using primitive types like `int`, `string` etc.

```C#
public class User
{
	public UserId Id { get; set; }
	public Nickname Nickname { get; set; }
}

public class UserDto
{
	public int Id { get; set; }
	public string Nickname { get; set; }
}
```

I want to use AgileMapper to map from `User` to `UserDto`. Unfortunately, I haven't found a way to do that. Here is the interesting part of the plan `Mapper.GetPlanFor<User>().ToANew<UserDto>()`:

```C#
var userDto = new UserDto();
// No data source for Id
userDto.Nickname = (string)uToUdData.Source.Nickname;
```

The `Nickname` conversion almost miraculously works without any additional configuration. On the other hand it seems that AgileMapper has decided to not map `Id`. Explicitly mapping `Id` using

```C#
Mapper.WhenMapping.From<User>().To<UserDto>().Map((source, target) => source.Id).To(target => target.Id);
```

throws an exception because AgileMapper is unable to convert `UserId` to an `int`. Adding a way to convert from `UserId` to `int`

```C#
Mapper.WhenMapping.From<UserId>().To<int>().CreateInstancesUsing(mappingData => (int) mappingData.Source);
```

seems to be ignored and the plan `Mapper.GetPlanFor<UserId>().Over<int>()` does nothing:

```C#
return uiToIData.Target;
```

I haven't found a way for me to configure how value types are converted. The built-in support for strings happens to work great with my code because my string-based types can be cast to `string`. However, a more general approach to be able to control this would make AgileMapper useful in projects that have a need to map value types. Another example is converting a GUID to a byte array with 16 elements. Can Agile Mapper be configured to perform this conversion?

**Comment by SteveWilkes** [comment]
Hi!

I think we'll have a couple of ways around this, yeah.

`Nickname` -> `string` works because AgileMapper knows to look for and use user-defined type conversion operators, and the same should work with `UserId` -> `int`, so I'll look at why that isn't working. Once that's fixed this should work without configuration.

Currently, v1.1 Preview 3 should support global configuration of value types using:

```C#
Mapper.WhenMapping
    .From<UserId>().To<int>()
    .Map((uid, id) => (int)uid)
    .ToTarget();
```

I should probably make `CreateInstancesUsing` for `int` throw a configuration exception as you can't create an 'instance' of a primitive - the method only lets you specify a value type to support struct configuration.

I'll look at this all soon - thanks for the feedback! :)

Steve

**Comment by SteveWilkes** [comment]
Turns out it was only checking for operators for to-string conversions - as of [v1.1-preview4](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.1.0-preview4) the `int` -> `UserId` and `UserId` -> `int` (and any other) operators will be picked up and used. 

Thanks for the feedback!

**Comment by SteveWilkes** [comment]
Automatic operator use and simple-Type `ToTarget()` configuration are included in [v1.1](https://www.nuget.org/packages/AgileObjects.AgileMapper).

Thanks! :)
